### ISSUE TITLE
Slow rendering of inputs with many paths

### ISSUE THREAD

**Issue by neinseg** [RazrFalcon/resvg#499]
resvg performs poorly on input files with many tens of thousands of individual paths. Attached are two files, slow.svg and fast.svg. slow.svg contains approximately 100k <path> elements with some style and transform set. Most of the paths have only two nodes. Converting it to PNG using resvg at 300 dpi takes approximately 15 minutes on a recent laptop AMD ryzen CPU.

fast.svg is a version of the file, where most of the 100k paths have been merged into a single path using some vim magic. It renders in less than one second everything else being the same.

Here is a plot of resvg's performance with an increasing number of input paths:

![out](https://user-images.githubusercontent.com/67436430/151708338-b056d00e-a845-411e-bcab-4af0e84c05dd.png)

This graph shows sub-linear performance beyond ~1000 input paths. Note that all tests have been run in parallel, so performance for shorter durations is probably under-estimated and for longer durations performance is likely over-estimated due to CPU congestion getting better the longer the test went on.

Test files:
[testfiles.zip](https://github.com/RazrFalcon/resvg/files/7966365/testfiles.zip)

**Comment by RazrFalcon** [comment]
This is the expected result. resvg doesn't support "batching", meaning that even when we have multiple paths with the same style, they still will be rendered separately. 

This is a known optimization possibility, but it would affect only very specific files.

PS: currently, resvg performance wasn't a priority at all. It's expected to be pretty slow. Correctness was and still is the main goal.

**Comment by FylmTM** [comment]
Hi!

We just got an interesting case in the wild, related to this. 
User uploaded SVG that is:
* 49 MB large 
* 49 million character long
* Contains 370 thousand occurrences of word "path"

Looks like that each pixel in the image is separate path.

This is absolutely bananas and we are taking steps now to prevent users from uploading SVG's like that.
In our particular scenario it sent our service into 100% CPU utilization until we killed it.

How people even come up with these? Urgh

**Comment by RazrFalcon** [comment]
SVG is way too complex which allows creation of such "zip bombs". There are at least two known way to freeze resvg. And both are relatively hard to fix/detect.

As for your particular issue, I'm not sure it can be fixed on the resvg side. resvg does have a hard limit on maximum number of elements, but we could make it configureable if needed.

**Comment by FylmTM** [comment]
Can you point to a code where hard limit is set? We can look into making a PR to make it configurable, if you are down for it.

**Comment by RazrFalcon** [comment]
Sure: https://github.com/RazrFalcon/resvg/blob/f991e0ba1f93ecb4ad79237966ff843ccf645b36/usvg/src/svgtree/parse.rs#L272

**Comment by RazrFalcon** [comment]
@yisibl this issue is not affected, sadly. I plan to address it in the next release.
